### ISSUE TITLE
TCK: Coupled operator vs rule 2.3

### ISSUE THREAD

**Issue by danielkec** [eclipse/microprofile-reactive-streams-operators#131]
Coupled operator [needs to signal cancel](https://download.eclipse.org/microprofile/microprofile-reactive-streams-operators-1.0/apidocs/org/eclipse/microprofile/reactive/streams/operators/ReactiveStreams.html#coupled-org.reactivestreams.Subscriber-org.reactivestreams.Publisher-) to parallel stream's upstream on its completion, but [rule 2.3](https://github.com/reactive-streams/reactive-streams-jvm#2.3) forbids direct call from **onComplete**  to **cancel**. Since the streams are actually parallel is the rule 2.3 and its tck test actually applicable? Same goes from onError->cancel call.

There was already discussion on https://gitter.im/eclipse/microprofile-reactive#

**Comment by Azquelt** [comment]
I don't think this violates 2.3 because you're not calling `cancel()` on the subscription from the publisher which sent the `onComplete()`.

I'm not sure what you mean by "parallel" stream. As I read it, the couple operator is a processor (and therefore has an upstream and a downstream), when you create it, you pass in a Subscriber `sub` and a Publisher `pub`. The couple then passes elements from `upstream` to `sub` and elements from `pub` to `downstream`.

In the scenario you're talking about, `upstream` has called `onComplete` and the couple has to
* call `onComplete` on `sub`
* call `cancel` on the subscription from `pub`
* call `onComplete` on `downstream`

I think rule 2.3 would only apply if you were to try to call `cancel` on the subscription received from `upstream`, which the spec does not require you to do.

**Comment by danielkec** [comment]
Yes but TCK test for rule 2.3 doesn't care about originating stream, it just looks for method called "onComplete" in call stack. Rule 2.3 should not be applied to coupled processor because its publisher and subscriber are not subscribed to each other right?

**Comment by cescoffier** [comment]
@danielkec Agreed. We need to fix that in the TCK, or if possible improve the detection.

**Comment by danielkec** [comment]
Related issue: reactive-streams/reactive-streams-jvm/issues/460

**Comment by olotenko** [comment]
I wonder how they are going to enforce rule 2.3 at all. Subscriptions are meant to be used concurrently. Plenty of rules specifically allowing `request` and `cancel` to be called synchronously, and references to possibility of `request` or `cancel` and `onNext` to be on different threads. All this implies it may well be during `onComplete`.

In other words, 2.3 is moot.

**Comment by akarnokd** [comment]
The TCK checked for any method named `onError` or `onComplete` in the stacktrace when a call was made to `request` or `cancel`, even though those originated from another thread or before the test itself called `onComplete` or `onError`.

Fix posted to the TCK: https://github.com/reactive-streams/reactive-streams-jvm/pull/483

**Comment by Emily-Jiang** [comment]
Waiting for Reactive Streams TCK release and then consume it.
